ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,393 @@
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_TRACKED_ALLOCATOR
+    }
+}
+
+impl AllocatorSelector for UnTracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_UNTRACKED_ALLOCATOR
+    }
+}
+
+#[derive(Debug)]
+pub struct Kva<M: AllocatorSelector + 'static> {
+    range: Range<Vaddr>,
+    phantom: PhantomData<M>,
+}
+
+impl<M: AllocatorSelector + 'static> Kva<M> {
+    pub fn new(size: usize) -> Self {
+        let allocator = M::select_allocator();
+        let range = allocator.alloc(size).unwrap();
+        Self {
+            range,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn start(&self) -> Vaddr {
+        self.range.start
+    }
+
+    pub fn end(&self) -> Vaddr {
+        self.range.end
+    }
+
+    pub fn range(&self) -> Range<Vaddr> {
+        self.range.start..self.range.end
+    }
+
+    fn query_page(&self, addr: Vaddr) -> PageTableItem {
+        assert!(self.start() <= addr && self.end() >= addr);
+        let start = addr.align_down(PAGE_SIZE);
+        let vaddr = start..start + PAGE_SIZE;
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let mut cursor = page_table.cursor(&vaddr).unwrap();
+        cursor.query().unwrap()
+    }
+}
+
+impl Kva<Tracked> {
+    /// Maps pages into the kernel virtual area.
+    pub fn map_pages<T: PageMeta>(
+        &mut self,
+        range: Range<Vaddr>,
+        pages: impl Iterator<Item = Page<T>>,
+        prop: PageProperty,
+    ) {
+        assert!(self.start() <= range.start && self.end() >= range.end);
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let mut cursor = page_table.cursor_mut(&range).unwrap();
+        let flusher = TlbFlusher::new(CpuSet::new_full(), disable_preempt());
+        let mut va = self.start();
+        for page in pages.into_iter() {
+            if let Some(old) = unsafe { cursor.map(page.into(), prop) } {
+                flusher.issue_tlb_flush_with(TlbFlushOp::Address(va), old);
+                flusher.dispatch_tlb_flush();
+            }
+            va += PAGE_SIZE;
+        }
+        flusher.issue_tlb_flush(TlbFlushOp::Range(range));
+        flusher.dispatch_tlb_flush();
+    }
+
+    /// Gets the mapped tracked page.
+    /// This function returns None in the following cases:
+    /// * If the address is not mapped (`NotMapped`).
+    /// * If the address is mapped to a `MappedUntracked` page.
+    pub fn get_page(&self, addr: Vaddr) -> Option<DynPage> {
+        let query_result = self.query_page(addr);
+        match query_result {
+            PageTableItem::Mapped {
+                va: _,
+                page,
+                prop: _,
+            } => Some(page),
+            _ => None,
+        }
+    }
+}
+
+impl Kva<UnTracked> {
+    /// Maps untracked pages into the kernel virtual area.
+    /// `pa_range.start` and `pa_range.end` should be aligned to PAGE_SIZE.
+    pub fn map_untracked_pages(
```

Comment:
This function should be `unsafe`. As the physical address is provided without guarantee, the caller must ensure that they are valid.

However, the `Kva<Tracked>` counterpart is safe.